network/wsNetwork.go

```diff
@@ -1488,14 +1544,44 @@ func (wn *WebsocketNetwork) tryConnect(addr, gossipAddr string) {
 	requestHeader.Set(InstanceNameHeader, myInstanceName)
 	conn, response, err := websocketDialer.DialContext(wn.ctx, gossipAddr, requestHeader)
 	if err != nil {
-		wn.log.Warnf("ws connect(%s) fail: %s", gossipAddr, err)
+		if err == websocket.ErrBadHandshake {
+			// reading here from ioutil is safe only because it came from DialContext above, which alredy finsihed reading all the data from the network
+			// and placed it all in a ioutil.NopCloser reader.
+			bodyBytes, _ := ioutil.ReadAll(response.Body)
+			errString := string(bodyBytes)
+
+			// we're gurenteed to have a valid response object.
+			switch response.StatusCode {
+			case http.StatusPreconditionFailed:
+				wn.log.Warnf("ws connect(%s) fail - bad handshake, precondition failed : %s error : '%s'", gossipAddr, errString)
+			case http.StatusLoopDetected:
+				wn.log.Infof("ws connect(%s) aborted due to connecting to self", gossipAddr)
+			default:
+				wn.log.Warnf("ws connect(%s) fail - bad handshake, Status code = %d, Headers = %#v, Body = %s", gossipAddr, response.StatusCode, response.Header, errString)
+			}
+		} else {
+			wn.log.Warnf("ws connect(%s) fail: %s", gossipAddr, err)
+		}
 		return
 	}
 	// no need to test the response.StatusCode since we know it's going to be http.StatusSwitchingProtocols, as it's already being tested inside websocketDialer.DialContext.
-	ok, otherTelemetryGUID, _, _ := wn.checkHeaders(response.Header, gossipAddr, nil)
-	if !ok {
+	// checking the headers here is abit redundent; the server has already verified that the headers match. But we will need this in the future -
+	// once our server would support multiple protocols, we would need to verify here that we use the correct protocol, out of the "proposed" protocols we have provided in the
+	// request headers.
+	otherTelemetryGUID, _, _, headersErr := wn.checkHeaders(response.Header, gossipAddr, nil)
+	if headersErr != nil {
 		return
 	}
+	otherGenesisID := response.Header.Get(GenesisHeader)
```

Comment:
Any reason this isn't inside checkHeaders()?

Reply:
Yes; it's being checked differently on the client vs. server.
On the server, we check the URI; on the client we check the response header.

Reply:
Not required but maybe having context-specific check functions would be clearer (eg checkOugoingHeaders() and checkIncomingHeaders())

Reply:
Agree; the existing code need to be slightly refactored to make it bit more structured.